geo/src/algorithm/interior_point.rs

```diff
@@ -0,0 +1,787 @@
+        if relation.is_intersects() {
+            return Some((
+                midpoint,
+                if relation.is_contains() {
+                    segment_length
+                } else {
+                    // if our point is on the boundary, it must be because this is a zero-area
+                    // polygon, so if we're being called from a multipolygon context, we want this
+                    // option to be down-ranked as compared to other polygons that might have
+                    // non-zero area
+                    T::zero()
+                },
+            ));
+        }
+    }
+    None
+}
+
+impl<T> InteriorPoint for Polygon<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    fn interior_point(&self) -> Self::Output {
+        polygon_interior_point_with_segment_length(self).map(|(point, _length)| point)
+    }
+}
+
+impl<T> InteriorPoint for MultiPolygon<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    fn interior_point(&self) -> Self::Output {
+        let segments = self
+            .iter()
+            .filter_map(polygon_interior_point_with_segment_length);
+        segments
+            .min_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(Ordering::Less))
+            .map(|(point, _length)| point)
+    }
+}
+
+impl<T> InteriorPoint for Rect<T>
+where
+    T: GeoFloat,
+{
+    type Output = Point<T>;
+
+    fn interior_point(&self) -> Self::Output {
+        self.center().into()
+    }
+}
+
+impl<T> InteriorPoint for Point<T>
+where
+    T: GeoFloat,
+{
+    type Output = Point<T>;
+
+    fn interior_point(&self) -> Self::Output {
+        *self
+    }
+}
+
+///
+/// ```
+/// use geo::InteriorPoint;
+/// use geo::{MultiPoint, Point};
+///
+/// let empty: Vec<Point> = Vec::new();
+/// let empty_multi_points: MultiPoint<_> = empty.into();
+/// assert_eq!(empty_multi_points.interior_point(), None);
+///
+/// let points: MultiPoint<_> = vec![(5., 1.), (1., 3.), (3., 2.)].into();
+/// assert_eq!(points.interior_point(), Some(Point::new(3., 2.)));
+/// ```
+impl<T> InteriorPoint for MultiPoint<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    fn interior_point(&self) -> Self::Output {
+        if let Some(centroid) = self.centroid() {
+            self.iter()
+                .map(|pt| (pt, pt.euclidean_distance(&centroid)))
+                .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Less))
+                .map(|(pt, _distance)| *pt)
+        } else {
+            None
+        }
+    }
+}
+
+impl<T> InteriorPoint for Geometry<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    crate::geometry_delegate_impl! {
+        fn interior_point(&self) -> Self::Output;
+    }
+}
+
+impl<T> InteriorPoint for GeometryCollection<T>
+where
+    T: GeoFloat,
+{
+    type Output = Option<Point<T>>;
+
+    fn interior_point(&self) -> Self::Output {
+        if let Some(centroid) = self.centroid() {
+            self.iter()
```

Comment:
According to JTS's documention, for mixed geometry collections JTS will return a point within the geometry with the highest dimensionality.

e.g. given `GEOMETRYCOLLECTION(POINT(...), POLYGON(...))`

The given point will always be within the Polygon (degenerate geometries perhaps notwithstanding).

I can imagine why that's useful. Is there a reason you diverged there?

We do have a `HasDimensions` which might be useful in such an implementation.

Reply:
Good call, I just didn't catch this intent on the JTS side. Updated to adopt this behavior, and added a test to capture it (`mixed_collection_test`).